// gatsby-plugin-utils/dist/test-plugin-options-schema.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.testPluginOptionsSchema = testPluginOptionsSchema;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _joi = require("./joi");

var _validate = require("./validate");

function testPluginOptionsSchema(_x, _x2) {
  return _testPluginOptionsSchema.apply(this, arguments);
}

function _testPluginOptionsSchema() {
  _testPluginOptionsSchema = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(pluginSchemaFunction, pluginOptions) {
    var pluginSchema, _warning$details$map, _warning$details, _yield$validateOption, warning, warnings, _e$details$map, _e$details, errors;

    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            pluginSchema = pluginSchemaFunction({
              Joi: _joi.Joi.extend(function (joi) {
                return {
                  type: "subPlugins",
                  base: joi.array().items(joi.alternatives(joi.string(), joi.object({
                    resolve: _joi.Joi.string(),
                    options: _joi.Joi.object({}).unknown(true)
                  }))).custom(function (arrayValue) {
                    return arrayValue.map(function (value) {
                      if (typeof value === "string") {
                        value = {
                          resolve: value
                        };
                      }

                      return value;
                    });
                  }, "Gatsby specific subplugin validation").default([])
                };
              })
            });
            _context.prev = 1;
            _context.next = 4;
            return (0, _validate.validateOptionsSchema)(pluginSchema, pluginOptions);

          case 4:
            _yield$validateOption = _context.sent;
            warning = _yield$validateOption.warning;
            warnings = (_warning$details$map = warning === null || warning === void 0 ? void 0 : (_warning$details = warning.details) === null || _warning$details === void 0 ? void 0 : _warning$details.map(function (detail) {
              return detail.message;
            })) !== null && _warning$details$map !== void 0 ? _warning$details$map : [];

            if (!((warnings === null || warnings === void 0 ? void 0 : warnings.length) > 0)) {
              _context.next = 9;
              break;
            }

            return _context.abrupt("return", {
              isValid: true,
              errors: [],
              hasWarnings: true,
              warnings: warnings
            });

          case 9:
            _context.next = 15;
            break;

          case 11:
            _context.prev = 11;
            _context.t0 = _context["catch"](1);
            errors = (_e$details$map = _context.t0 === null || _context.t0 === void 0 ? void 0 : (_e$details = _context.t0.details) === null || _e$details === void 0 ? void 0 : _e$details.map(function (detail) {
              return detail.message;
            })) !== null && _e$details$map !== void 0 ? _e$details$map : [];
            return _context.abrupt("return", {
              isValid: false,
              errors: errors,
              hasWarnings: false,
              warnings: []
            });

          case 15:
            return _context.abrupt("return", {
              isValid: true,
              errors: [],
              hasWarnings: false,
              warnings: []
            });

          case 16:
          case "end":
            return _context.stop();
        }
      }
    }, _callee, null, [[1, 11]]);
  }));
  return _testPluginOptionsSchema.apply(this, arguments);
}
