// gatsby-plugin-utils/dist/validate.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.validateOptionsSchema = validateOptionsSchema;

var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _asyncToGenerator2 = _interopRequireDefault(require("@babel/runtime/helpers/asyncToGenerator"));

var _joi = require("./joi");

var validationOptions = {
  // Show all errors at once, rather than only the first one every time
  abortEarly: false,
  cache: true
};

function validateOptionsSchema(_x, _x2, _x3) {
  return _validateOptionsSchema.apply(this, arguments);
}

function _validateOptionsSchema() {
  _validateOptionsSchema = (0, _asyncToGenerator2.default)( /*#__PURE__*/_regenerator.default.mark(function _callee(pluginSchema, pluginOptions, options) {
    var _options, validateExternalRules, returnWarnings, warnOnUnknownSchema;

    return _regenerator.default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            if (options === void 0) {
              options = {
                validateExternalRules: true,
                returnWarnings: true
              };
            }

            _options = options, validateExternalRules = _options.validateExternalRules, returnWarnings = _options.returnWarnings;
            warnOnUnknownSchema = pluginSchema.pattern(/.*/, _joi.Joi.any().warning("any.unknown"));
            _context.next = 5;
            return warnOnUnknownSchema.validateAsync(pluginOptions, (0, _extends2.default)({}, validationOptions, {
              externals: validateExternalRules,
              warnings: returnWarnings
            }));

          case 5:
            return _context.abrupt("return", _context.sent);

          case 6:
          case "end":
            return _context.stop();
        }
      }
    }, _callee);
  }));
  return _validateOptionsSchema.apply(this, arguments);
}
