// gatsby-plugin-manifest/gatsby-ssr.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var React = _interopRequireWildcard(require("react"));

var _gatsby = require("gatsby");

var _common = require("./common.js");

var _getManifestPathname = _interopRequireDefault(require("./get-manifest-pathname"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

// TODO: remove for v3
const withPrefix = _gatsby.withAssetPrefix || _gatsby.withPrefix;

exports.onRenderBody = ({
  setHeadComponents,
  pathname = `/`
}, {
  localize,
  legacy,
  cache_busting_mode: cacheBusting,
  cacheDigest,
  icon,
  icons: pluginIcons,
  include_favicon: insertFaviconLinkTag,
  theme_color_in_head: insertMetaTag,
  theme_color: themeColor,
  crossOrigin = `anonymous`
}) => {
  // We use this to build a final array to pass as the argument to setHeadComponents at the end of onRenderBody.
  const headComponents = [];
  const srcIconExists = !!icon;
  const icons = pluginIcons || _common.defaultIcons;
  const manifestFileName = (0, _getManifestPathname.default)(pathname, localize); // If icons were generated, also add a favicon link.

  if (srcIconExists) {
    if (insertFaviconLinkTag) {
      _common.favicons.forEach(favicon => {
        headComponents.push( /*#__PURE__*/React.createElement("link", {
          key: `gatsby-plugin-manifest-icon-link-png`,
          rel: "icon",
          href: withPrefix((0, _common.addDigestToPath)(favicon.src, cacheDigest, cacheBusting)),
          type: "image/png"
        }));
      });

      if (icon !== null && icon !== void 0 && icon.endsWith(`.svg`)) {
        headComponents.push( /*#__PURE__*/React.createElement("link", {
          key: `gatsby-plugin-manifest-icon-link-svg`,
          rel: "icon",
          href: withPrefix((0, _common.addDigestToPath)(`favicon.svg`, cacheDigest, cacheBusting)),
          type: "image/svg+xml"
        }));
      }
    }
  } // Add manifest link tag.


  headComponents.push( /*#__PURE__*/React.createElement("link", {
    key: `gatsby-plugin-manifest-link`,
    rel: "manifest",
    href: (0, _gatsby.withPrefix)(`/${manifestFileName}`),
    crossOrigin: crossOrigin
  })); // The user has an option to opt out of the theme_color meta tag being inserted into the head.

  if (themeColor && insertMetaTag) {
    headComponents.push( /*#__PURE__*/React.createElement("meta", {
      key: `gatsby-plugin-manifest-meta`,
      name: "theme-color",
      content: themeColor
    }));
  }

  if (legacy) {
    icons.forEach(icon => {
      headComponents.push( /*#__PURE__*/React.createElement("link", {
        key: `gatsby-plugin-manifest-apple-touch-icon-${icon.sizes}`,
        rel: "apple-touch-icon",
        sizes: icon.sizes,
        href: withPrefix((0, _common.addDigestToPath)(icon.src, cacheDigest, srcIconExists ? cacheBusting : `none`))
      }));
    });
  }

  setHeadComponents(headComponents);
  return true;
};
