// gatsby-core-utils/dist/index.js
"use strict";

exports.__esModule = true;
var _exportNames = {
  createContentDigest: true,
  isNodeInternalModulePath: true,
  joinPath: true,
  slash: true,
  cpuCoreCount: true,
  urlResolve: true,
  getCIName: true,
  isCI: true,
  createRequireFromPath: true,
  getConfigStore: true,
  getGatsbyVersion: true,
  getTermProgram: true,
  fetchRemoteFile: true,
  IFetchRemoteFileOptions: true,
  isTruthy: true,
  getMatchPath: true,
  listPlugins: true,
  createFilePath: true,
  readConfigFile: true,
  getConfigPath: true,
  lock: true,
  uuid: true
};
exports.uuid = exports.lock = exports.getConfigPath = exports.readConfigFile = exports.createFilePath = exports.listPlugins = exports.getMatchPath = exports.isTruthy = exports.IFetchRemoteFileOptions = exports.fetchRemoteFile = exports.getTermProgram = exports.getGatsbyVersion = exports.getConfigStore = exports.createRequireFromPath = exports.isCI = exports.getCIName = exports.urlResolve = exports.cpuCoreCount = exports.slash = exports.joinPath = exports.isNodeInternalModulePath = exports.createContentDigest = void 0;

var _createContentDigest = require("./create-content-digest");

exports.createContentDigest = _createContentDigest.createContentDigest;

var _path = require("./path");

exports.isNodeInternalModulePath = _path.isNodeInternalModulePath;
exports.joinPath = _path.joinPath;
exports.slash = _path.slash;

var _cpuCoreCount = require("./cpu-core-count");

exports.cpuCoreCount = _cpuCoreCount.cpuCoreCount;

var _url = require("./url");

exports.urlResolve = _url.urlResolve;

var _ci = require("./ci");

exports.getCIName = _ci.getCIName;
exports.isCI = _ci.isCI;

var _createRequireFromPath = require("./create-require-from-path");

exports.createRequireFromPath = _createRequireFromPath.createRequireFromPath;

var _getConfigStore = require("./get-config-store");

exports.getConfigStore = _getConfigStore.getConfigStore;

var _getGatsbyVersion = require("./get-gatsby-version");

exports.getGatsbyVersion = _getGatsbyVersion.getGatsbyVersion;

var _getTermProgram = require("./get-term-program");

exports.getTermProgram = _getTermProgram.getTermProgram;

var _fetchRemoteFile = require("./fetch-remote-file");

exports.fetchRemoteFile = _fetchRemoteFile.fetchRemoteFile;
exports.IFetchRemoteFileOptions = _fetchRemoteFile.IFetchRemoteFileOptions;

var _isTruthy = require("./is-truthy");

exports.isTruthy = _isTruthy.isTruthy;

var _uuid = _interopRequireWildcard(require("./uuid"));

exports.uuid = _uuid;

var _matchPath = require("./match-path");

exports.getMatchPath = _matchPath.getMatchPath;

var _serviceLock = require("./service-lock");

Object.keys(_serviceLock).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _serviceLock[key]) return;
  exports[key] = _serviceLock[key];
});

var _siteMetadata = require("./site-metadata");

Object.keys(_siteMetadata).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _siteMetadata[key]) return;
  exports[key] = _siteMetadata[key];
});

var _pageData = require("./page-data");

Object.keys(_pageData).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _pageData[key]) return;
  exports[key] = _pageData[key];
});

var _pageHtml = require("./page-html");

Object.keys(_pageHtml).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _pageHtml[key]) return;
  exports[key] = _pageHtml[key];
});

var _listPlugins = require("./list-plugins");

exports.listPlugins = _listPlugins.listPlugins;

var _filenameUtils = require("./filename-utils");

exports.createFilePath = _filenameUtils.createFilePath;

var _utils = require("./utils");

exports.readConfigFile = _utils.readConfigFile;
exports.getConfigPath = _utils.getConfigPath;

var _lock = require("./lock");

exports.lock = _lock.lock;

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function (nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
